---
title: "Class 14: Genome Informatics II"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
metadata <- read.csv("airway_metadata.csv", stringsAsFactors = FALSE)
counts <- read.csv("airway_scaledcounts.csv", stringsAsFactors = FALSE)
```

```{r}
head(metadata)
head(counts)
```
Q1: How many genes are in this dataset?
```{r}
nrow(counts)
```

Q2: How many 'control' cell lines do we have? 
```{r}
dim(metadata)
metadata
```
Let's check if the cols of `counts` is the same as the id column in `metadata`
```{r}
colnames(counts)
```

```{r}
metadata$id
```

```{r}
all(colnames(counts)[-1] == metadata$id)
```

Separate control and treated data
```{r}
control <- metadata[metadata[,"dex"]=="control",]
control
```




Now we want to get the cols in counts with these `control$id` colnames
```{r}
control.mean <- rowSums(counts[,control$id])/length(control$id)
names(control.mean) <- counts$ensgene
head(control.mean)
```

Do the same for the treated samples
```{r}
treated <- metadata[metadata$dex == "treated",]
treated
treated.mean <- rowSums(counts[,treated$id])/length(treated$id)
names(treated.mean) <- counts$ensgene
head(treated.mean)
```


Let's put these together into a new data.frame called `meancounts`
```{r}
meancounts <- data.frame(control.mean, treated.mean)
head(meancounts)
```
```{r}
plot(meancounts, xlab = "Control", ylab = "Treated")
```




Plot on a log scale for both axes.
```{r}
plot(meancounts, xlab = "Control", ylab = "Treated", log="xy")
```

```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```


The NaN is returned when you divide by zero and try to take the log. The -Inf is returned when you try to take the log of zero. It turns out that there are a lot of genes with zero expression. Let’s filter our data to remove these genes.

```{r}
# Identify the rows (i.e. genes) with zero values
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)

# Remove these rows as we can't really say much about them.
to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```
Q8: How many genes are upregulated/downregulated at the greater than 2 fc level?
```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```

```{r}
#mycounts[up.ind,]
sum(up.ind)
```


```{r}
#mycounts[down.ind,]
sum(down.ind)
```



## 4. Adding annotation data

```{r}
anno <- read.csv("annotables_grch38.csv")
```


We will use the **merge()** function to add annotation data to our `mycounts`

Q10. From consulting the help page for the merge() function can you set the by.x and by.y arguments appropriately to annotate our mycounts data.frame with all the available annotation data in your anno data.frame?
```{r}
mycounts.anno <- merge(x=mycounts, y=anno, by.x="row.names", by.y = "ensgene" )
```

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
```

```{r}
columns(org.Hs.eg.db)
```


```{r}
mycounts$symbol <- mapIds(org.Hs.eg.db, keys=row.names(mycounts), keytype="ENSEMBL", column="SYMBOL", multiVals="first")
```


Q11. Run the mapIds() function two more times to add the Entrez ID and UniProt accession as new columns called mycounts$entrez and mycounts$uniprot.
```{r}
mycounts$entrez <- mapIds(org.Hs.eg.db, keys=row.names(mycounts), keytype="ENSEMBL", column="ENTREZID", multiVals="first")
mycounts$uniprot <- mapIds(org.Hs.eg.db, keys=row.names(mycounts), keytype="ENSEMBL", column="UNIPROT", multiVals="first")
```

```{r}
head(mycounts)
```

Q12. Examine your annotated results for those genes with a log2(FoldChange) of greater than 2 (or less than -2 if you prefer) with the View( mycounts[up.ind,] ) function. What do you notice? Would you trust these results? Why or why not?
```{r}
View(mycounts[up.ind,])
```



## 5. DESeq2 Analysis


Let's do this the right way. DESeq2 is an R package for analyzing count-based NGS data like RNA-seq. It is available from Bioconductor. 


```{r}
library("DESeq2")
citation("DESeq2")
```

Build our deseq dataset object need for running deseq analysis
```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
dds
```


```{r}
dds <- DESeq(dds)
```

Get our results
```{r}
res <- results(dds)
#as.data.frame(res)
```


```{r}
summary(res)
```


```{r}
res05 <- results(dds, alpha=0.05)
summary(res05)
```

## 6. Data Visualization
Typically these plots shows the log fold change on the X-axis, and the −log10
 of the p-value on the Y-axis (the more significant the p-value, the larger the −log10
 of that value will be). A very dull (i.e. non colored and labeled) version can be created with a quick call to plot() like so:
 
 
 higher up means more significant
 0 on the x-axis means exactly the same (no change in expression)
 negaive values --> the genes are downregulated in the treated 
 

```{r}
plot(res$log2FoldChange,  -log(res$padj), 
      xlab="Log2(FoldChange)",
      ylab="-Log(P-value)")
```
 
 

```{r}
plot( res$log2FoldChange,  -log(res$padj), 
      xlab="Log2(FoldChange)",
      ylab="-Log(P-value)")
abline(v=c(-2,2), col="darkgray", lty=2)
abline(h=-log(0.05), col="darkgray", lty=2)
```
 
 
 
 
```{r}
# Setup our custom point color vector 
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2 ]  <- "red" 

inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

# Volcano plot with custom colors 
plot( res$log2FoldChange,  -log(res$padj), 
 col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)" )

# Cut-off lines
abline(v=c(-2,2), col="gray", lty=2)
abline(h=-log(0.1), col="gray", lty=2)
```
 
Another way to plot this is with the Enhanced Volcano function

But first lets put proper gene names on our results (i.e. gene `symbols`)

```{r}
x <- as.data.frame(res)
x$symbol <- mapIds(org.Hs.eg.db, 
                   keys=row.names(x),
                   keytype="ENSEMBL",
                   column="SYMBOL",
                   multiVals="first")
```


```{r}
library(EnhancedVolcano)
```



```{r}
EnhancedVolcano(x,
    lab = x$symbol,
    x = 'log2FoldChange',
    y = 'pvalue')
```


